lean/monad-stacks.lean

```diff
@@ -1,33 +1,250 @@
 /-!
 # Monad Stacks
+Before you read this: This chapter does not attempt to introduce the concept
+of a monad in general, it assume this as a given and explains the monads
+and constructs on top of them that are essential to Lean meta programming.
+If you don't know what a monad is already you can read (TODO: Link to monad tutorial).
 
-# Monad transformers and Mondas relevant for meta
-* ReaderT
-* StateRefT
-
-# Monad extensions
-* MonadLift
-* MonadState
-* MonadExcept
-* MonadReader
-* MonadControl
-* MonadFinally
-* MonadFunctor
-* MonadStateOf
-* MonadControlT
-* MonadEnv
-* MonadRef
-* MonadMCtx
-* MonadLCtx
-* MonadTrace
-* MonadCache
-* MonadError
-* MonadWithReader
-* MonadOptions
-* MonadRecDepth
-* MonadBacktrack
-
-* MonadError
-* ...
-* `liftM`
+## Monad Transformers
+Quite often in functional programming with monads one wants to have the power
+of more than one monad available. Specifically in Lean meta programming it
+is a very common pattern that we have some sort of read only input, an
+execution context, available to us, as well as the ability to operate on
+a mutatable state. Both of these issues have monads that can solve them
+seperately already, namely the `Reader` monad for read only input and the
+`StateM` monad for the mutable state. Now we can try to just "stack" them
+ontop like this:`
+-/
+namespace Playground1
+abbrev Context := Nat
+abbrev State := String
+
+abbrev MyM (α : Type) := Reader Context (StateM State α)
+
+end Playground1
+/-!
+And this will indeed type check, however we have a very annoying issue:
+-/
+namespace Playground1
+
+#synth Monad MyM -- no monad instance :(
```

Comment:
What is this command for?

Reply:
`#synth` attempts to *synthesize* an instance of the type it is given and fails if it cannot do that.

Reply:
What happens if a monad instance cannot be synthesized?